### ISSUE TITLE
AOA USB connectivity stops working using the 4.7.1 Android proxy library.

### ISSUE THREAD

**Issue by hypatankar** [smartdevicelink/sdl_java_suite#924]
### Bug Report
Frequently, AOA USB connectivity stops working after 2-4 connect and disconnect events when using the 4.7.1 Android proxy library.

##### Reproduction Steps
1.	Build the Hello SDL sample application included with the SDL Android proxy library and install it on the Android mobile device.
2.	Stand up an instance of SDL core 5.0.0 on Linux or acquire hardware such as a Sync TDK that has AOA USB transport support.
3.	Physically connect the USB cable to the TDK or PC running SDL core.
4.    **Project Maintainers note: Set app as default app for USB connection when dialog displays**
5.	Take note that the Hello SDL sample app generally successfully registers with SDL core.
6.	Physically disconnect the USB cable from the TDK or PC.
7.	Repeat steps 4 and 5 more than three times, take note that many times after two to four connect and disconnect events the app will generally not continually register over the AOA USB transport.  

##### Expected Behavior
The HelloSDL app (and proxy library) should register every time a mobile device is connected with SDL core.

##### Observed Behavior
Frequently, HelloSDL registers only the first few times then doesn’t register. 

##### OS & Version Information
* Various Android versions including Android 9 and below.
* Sync TDK, SDL Core on Linux
* SDL Android Version is 4.7.1

##### Test Case, Sample Code, and / or Example App
N/A

**Comment by joeygrover** [comment]
Does it stop connecting all together? Or simply that connection did not connect and unplugging and plugging back in connects?

**Comment by joeygrover** [comment]
Is it the app not registering or is the actual transport not connecting?

**Comment by mrapitis** [comment]
It looks like the transport isn't getting a connection as the app registration request never occurs..... it should be fairly easy to repro as we used a variety of devices with basically the same results.

**Comment by joeygrover** [comment]
Can you attach logs of the events?

**Comment by mrapitis** [comment]
[logs.zip](https://github.com/smartdevicelink/sdl_android/files/2577626/logs.zip)

**Comment by joeygrover** [comment]
It seems like this is the exception or similar:

`java.io.FileNotFoundException: /jacoco.exec: open failed: EROFS (Read-only file system)`

Based on this https://github.com/commons-app/apps-android-commons/issues/1483 I'm wondering if you can try with a nondebug build and see if the issue remains.

**Comment by mrapitis** [comment]
The issue seems to still persist after creating / deploying a nondebug build as well... looks like it may not be the culprit.

**Comment by joeygrover** [comment]
On quick investigation I am seeing `UsbDeviceManagerJNI: could not open /dev/usb_accessory` error.

**Comment by joeygrover** [comment]
I've been trying a lot of different items but I'm starting to think this is an issue outside of the SDL library. The lower level calls seem to be failing at the native layer and there's no clear indication of why. So to be clear this is an issue with the USB accessory being opened, not an app registering. the `USBAccessoryAttachmentActivity` is never started so it's before the SDL library can begin.

Here are the logs for when it succeeds and when it fails.

### Success:

```
2018-11-21 13:42:33.278 750-1377/? I/android.hardware.usb@1.1-service.wahoo: partner added
2018-11-21 13:42:33.279 750-1377/? I/android.hardware.usb@1.1-service.wahoo: uevent received DEVTYPE=typec_partner
2018-11-21 13:42:33.283 750-1377/? I/android.hardware.usb@1.1-service.wahoo: port0
2018-11-21 13:42:33.285 750-1377/? I/android.hardware.usb@1.1-service.wahoo: connected:1 canChangeMode:1 canChagedata:0 canChangePower:0
2018-11-21 13:42:33.287 1112-23136/? I/UsbPortManager: ClientCallback: port0
2018-11-21 13:42:33.288 1112-1152/? I/UsbPortManager: USB port changed: port=UsbPort{id=port0, supportedModes=dual}, status=UsbPortStatus{connected=true, currentMode=ufp, currentPowerRole=sink, currentDataRole=device, supportedRoleCombinations=[source:host, sink:device]}, canChangeMode=true, canChangePowerRole=false, canChangeDataRole=false
2018-11-21 13:42:33.300 1543-1543/? D/ModemService: Received: android.hardware.usb.action.USB_PORT_CHANGED
2018-11-21 13:42:33.301 1543-1543/? D/ModemService: USB Port Status: UsbPortStatus{connected=true, currentMode=ufp, currentPowerRole=sink, currentDataRole=device, supportedRoleCombinations=[source:host, sink:device]}
2018-11-21 13:42:33.305 1543-1543/? D/ModemService: ModemService started
2018-11-21 13:42:33.306 1543-1636/? D/ModemService: CheckUpdateRunnable started
2018-11-21 13:42:33.306 1543-1636/? D/ModemService: Tuner by USB: enabled
2018-11-21 13:42:33.308 1543-1636/? D/ModemService: Sending tuner update
2018-11-21 13:42:33.308 1543-1636/? D/ModemService: Tuner value: 5
2018-11-21 13:42:33.308 1543-1636/? V/QMI_OEMHOOK: createPayload: serviceId= 4 messageId= 32
2018-11-21 13:42:33.309 1543-1636/? V/QMI_OEMHOOK: val=5
2018-11-21 13:42:33.309 1543-1636/? V/QMI_OEMHOOK: Intermediate buf in QmiOemHook sendQmiMessage Sync or Async = 0105000105000000
2018-11-21 13:42:33.309 1543-1636/? V/QMI_OEMHOOK: Byte buf in QmiOemHook createPayload = java.nio.HeapByteBuffer[pos=20 lim=20 cap=20]
2018-11-21 13:42:33.311 1668-1668/? D/QcrilMsgTunnelIfaceManager: handleMessage what = 1
2018-11-21 13:42:33.313 1668-1668/? I/QcrilOemhookMsgTunnel: [0]invokeOemRilRequestRaw: serial=1198 length=36
2018-11-21 13:42:33.325 1668-1702/? I/QcrilOemhookMsgTunnel: [0]processOemHookResponse serial no: 1198 error: 0length=0
2018-11-21 13:42:33.326 1668-1668/? D/QcrilMsgTunnelIfaceManager: handleMessage what = 2
2018-11-21 13:42:33.327 1543-1636/? V/QMI_OEMHOOK: receive responseData = null message=null responseType= IS_SYNC_RESPONSE
2018-11-21 13:42:33.327 1543-1636/? V/QMI_OEMHOOK: payload is null
2018-11-21 13:42:33.327 1543-1636/? D/ModemService: SAR by USB Device: disabled
2018-11-21 13:42:33.415 1112-1152/? D/UsbDeviceManager: push notification:USB charging this device
2018-11-21 13:42:33.465 1112-1152/? D/UsbDeviceManager: applyOemOverride usbfunctions=accessory,adb bootmode=unknown
2018-11-21 13:42:33.465 1112-1152/? I/UsbDeviceManager: Setting USB config to accessory,adb
2018-11-21 13:42:33.528 4036-4037/? I/adbd: initializing functionfs
2018-11-21 13:42:33.528 4036-4037/? I/adbd: opening control endpoint /dev/usb-ffs/adb/ep0
2018-11-21 13:42:33.923 1112-1152/? D/UsbDeviceManager: entering USB accessory mode: UsbAccessory[mManufacturer=SDL, mModel=Core, mDescription=null, mVersion=1.0, mUri=null, mSerial=null]
2018-11-21 13:42:33.939 4036-4037/? I/adbd: functionfs successfully initialized
2018-11-21 13:42:33.939 4036-4037/? I/adbd: registering usb transport
```

### Fail
```
2018-11-21 13:42:47.015 750-1377/? I/android.hardware.usb@1.1-service.wahoo: partner added
2018-11-21 13:42:47.016 750-1377/? I/android.hardware.usb@1.1-service.wahoo: uevent received DEVTYPE=typec_partner
2018-11-21 13:42:47.019 750-1377/? I/android.hardware.usb@1.1-service.wahoo: port0
2018-11-21 13:42:47.021 750-1377/? I/android.hardware.usb@1.1-service.wahoo: connected:1 canChangeMode:1 canChagedata:0 canChangePower:0
2018-11-21 13:42:47.023 1112-23136/? I/UsbPortManager: ClientCallback: port0
2018-11-21 13:42:47.024 1112-1152/? I/UsbPortManager: USB port changed: port=UsbPort{id=port0, supportedModes=dual}, status=UsbPortStatus{connected=true, currentMode=ufp, currentPowerRole=sink, currentDataRole=device, supportedRoleCombinations=[source:host, sink:device]}, canChangeMode=true, canChangePowerRole=false, canChangeDataRole=false
2018-11-21 13:42:47.037 1543-1543/? D/ModemService: Received: android.hardware.usb.action.USB_PORT_CHANGED
2018-11-21 13:42:47.038 1543-1543/? D/ModemService: USB Port Status: UsbPortStatus{connected=true, currentMode=ufp, currentPowerRole=sink, currentDataRole=device, supportedRoleCombinations=[source:host, sink:device]}
2018-11-21 13:42:47.042 1543-1543/? D/ModemService: ModemService started
2018-11-21 13:42:47.043 1543-1636/? D/ModemService: CheckUpdateRunnable started
2018-11-21 13:42:47.043 1543-1636/? D/ModemService: Tuner by USB: enabled
2018-11-21 13:42:47.048 1543-1636/? D/ModemService: Sending tuner update
2018-11-21 13:42:47.048 1543-1636/? D/ModemService: Tuner value: 5
2018-11-21 13:42:47.049 1543-1636/? V/QMI_OEMHOOK: createPayload: serviceId= 4 messageId= 32
2018-11-21 13:42:47.049 1543-1636/? V/QMI_OEMHOOK: val=5
2018-11-21 13:42:47.050 1543-1636/? V/QMI_OEMHOOK: Intermediate buf in QmiOemHook sendQmiMessage Sync or Async = 0105000105000000
2018-11-21 13:42:47.050 1543-1636/? V/QMI_OEMHOOK: Byte buf in QmiOemHook createPayload = java.nio.HeapByteBuffer[pos=20 lim=20 cap=20]
2018-11-21 13:42:47.052 1668-1668/? D/QcrilMsgTunnelIfaceManager: handleMessage what = 1
2018-11-21 13:42:47.055 1668-1668/? I/QcrilOemhookMsgTunnel: [0]invokeOemRilRequestRaw: serial=1200 length=36
2018-11-21 13:42:47.068 1668-1702/? I/QcrilOemhookMsgTunnel: [0]processOemHookResponse serial no: 1200 error: 0length=0
2018-11-21 13:42:47.069 1668-1668/? D/QcrilMsgTunnelIfaceManager: handleMessage what = 2
2018-11-21 13:42:47.075 1543-1636/? V/QMI_OEMHOOK: receive responseData = null message=null responseType= IS_SYNC_RESPONSE
2018-11-21 13:42:47.075 1543-1636/? V/QMI_OEMHOOK: payload is null
2018-11-21 13:42:47.075 1543-1636/? D/ModemService: SAR by USB Device: disabled
2018-11-21 13:42:47.107 1112-1152/? D/UsbDeviceManager: push notification:USB charging this device
2018-11-21 13:42:47.158 1112-1248/? E/UsbDeviceManagerJNI: could not open /dev/usb_accessory
2018-11-21 13:42:47.158 1112-1248/? E/UsbDeviceManagerJNI: could not open /dev/usb_accessory
```

**Comment by joeygrover** [comment]
I also observed that Android O displayed this behavior much more often and more unable to recover. After updating a phone to Android P, the error was much less frequent and even after happening the next physical connection would work as expected. It will be worthwhile to investigate frequency based on OS version. @mrapitis @hypatankar Can you perform these tests on various versions of the OS?

1. Connect USB
2. Observe connection was made, router service starts, and app registers
3. Disconnect USB
4. Reconnect USB
5. Observe if connection was made, router service starts, and app registers
6. Move back to step 3

Then take note if the connection is successful, how many times until failure, and if repeating the steps after a failure resulted in success or continued failure.

**Comment by mrapitis** [comment]
> I'm starting to think this is an issue outside of the SDL library. 

We do not believe this is likely an issue outside of the current version of the SDL library as we have built sample apps like HelloSDL utilizing the older version of the SDL Android library with the USB transport and we do not see this issue.  

We are working on compiling our test results with the current version of the SDL library with multiple Android devices and multiple versions of the Android OS (including O and P as requested).  We are planning to share these results today for the teams review.

**Comment by hypatankar** [comment]
I have attached some test results for review; testing was performed on a sync3 tdk. 

[usb-aoa-testing.pdf](https://github.com/smartdevicelink/sdl_android/files/2621139/usb-aoa-testing.pdf)

**Comment by joeygrover** [comment]
@hypatankar what build flavor of the hello sdl app were you using?

**Comment by hypatankar** [comment]
@joeygrover It was built with multi_high_bandwidth.

**Comment by joeygrover** [comment]
At this point I'm pretty much able to confirm the issue arises from opening the USB accessory and passing the `ParcelFileDescriptor` from one app to a router service. It seems there is some synchronization/race condition issues of when the PFD gets closed from one process to another, and this might be why the performance is different from phone to phone, OS to OS. This causes lock ups and error states for the native system to open the USB accessory. 

 I have a "fix" that essentially can prove this, however, the UX is pretty terrible. It would require the user to select an app to receive the USB intent (as it does today), then that app would bind to the router service and pass it the actual USB accessory, the router service would then check if it had permission to open it itself, if not it would ask the user for permission. The issue arises when the user selects one app to use USB then another dialog right after pops up with a different app requesting USB; this would be the app with the newest router service so it could be an app that doesn't even use USB. However, I believe this would also require apps to add the USB manifest items to their app as well. I don't find this acceptable as a solution and will continue to investigate alternatives but any help is welcome.

**Comment by joeygrover** [comment]
@mrapitis @hypatankar Please check #933 PR. There is still a lingering issue where apps are not properly ending their session, but it seems the USB connection is more reliable. It is not 100% more like 90%, but without introducing the previous comments multiple dialog scenario I'm not sure if 100% is a place we can truly get to.

Please pay attention to the testing details and ask any questions you have around that. Essentially if the system displays the App choice dialog for the USB attachment, then the `SdlRouterService` icon should come up each time. I will continue to investigate the incorrect session ending on the proxy side.

**Comment by joeygrover** [comment]
At this point the connection appears by all checkable means to be valid, but doesn't actually seem to be writing out to the device. I have confirmed the client library is creating the StartService packet and makes it way to the USB outputstream. From what I can tell, after the UsbProvider passes the PFD to the router service, a new PFD is essentially created with a different FileDescriptor that I believe is supposed to map to the USB FD. I have noticed this FD is the same value each time at the router service, however, the FD at the UsbProvider stage is normally different each connection. I'm guessing there is some mismatch between the FDs and the true connection.

I've added another branch that performs a wait before opening the USB Accessory that can be found [here](https://github.com/smartdevicelink/sdl_android/tree/bugfix/usb_enhacements_wait)

**Comment by mrapitis** [comment]
In testing we have noticed improved connectivity with the branches referenced when compared to the current master branch, however like previously discussed there are important cases where connectivity becomes entirely unusable.  A force close of the SDL enabled application from the mobile device or a reboot of the mobile device works around the issue (obviously not acceptable UX).  

During testing we also noticed something alarming, for each connection attempted / established over AOA, a new PacketWriteTaskMaster  thread is spawned while existing threads continue to be in the waiting state.  For example, we made 15 separate connections with a TDK and noticed 15 PacketWriteTaskMaster  threads.  Each connection increased the number of PacketWriteTaskMaster  threads by one.  I have attached a screenshot and a zip file (below) with mobile side thread report logs to provide further details.

![thread-view](https://user-images.githubusercontent.com/8443128/49377444-30638f00-f6d8-11e8-92b3-1b83f4e673d8.png)



[aoa-threads.zip](https://github.com/smartdevicelink/sdl_android/files/2639613/aoa-threads.zip)

**Comment by joeygrover** [comment]
Easy enough fix. There was a call to close the thread which set a stop flag, but since the thread was waiting it never ran through the conditional again to close itself. I added the alert call to wake it up. Also added a few other clean up items around that feature that were missing. However, while these were necessary changes they were not affecting the USB connectivity directly.

I have found a new interesting Android behavior. If you connect the app to the TDK via USB, see the dialog pop up, disconnect from the TDK and then accept the dialog, Android will allow the app to "open" the accessory, returns a real `ParcelFileDescriptor` that's attached to a "_valid_" `FileDescriptor`. This seems like a very poor issue with the OS itself, but might be another symptom of the same issue we're facing. The system will hand out the PFD even though it isn't actually valid....

**Comment by joeygrover** [comment]
@mrapitis I have confirmed the issue does not persist on the open source version of Core. I was able to connect/disconnect/reconnect 30 times in rapid succession with no faults and app registration each and every time. I asked @JackLivio to witness this and demonstrated the TDK failure during the same test. There might be an issue with how the QNX AOA transport assigns/requests file descriptors that would cause it to be unable to handle the new transport scheme. If we can find out where it is failing on the TDK we might be able to address what is causing it on the proxy side or come up with a workaround.

**Comment by mrapitis** [comment]
@joeygrover 
We were able to reproduce this issue with SDL core on linux using an LG-H820 (LG G5) running Android 7.0.  It does not appear to be a QNX specific issue....

**Comment by joeygrover** [comment]
@mrapitis Livio doesn't have any newer LG phones, is it possible to have Ford stop at the Livio office with some to perform testing? The Samsung devices would be helpful too. We are in the process of getting more devices, but as this is a critical issue this will be the fastest path.

**Comment by Tnnnguyen** [comment]
I had some mixed results using 2 phones using the steps to reproduce above.
Nexus 6, OS: Android 7.1.1: consistently connects.
Pixel XL, OS: Android 9: not consistent, will fail to connect randomly.
I have attached logs for each device.
[pixelXL.txt](https://github.com/smartdevicelink/sdl_android/files/2645659/pixelXL.txt)
[nexus.txt](https://github.com/smartdevicelink/sdl_android/files/2645660/nexus.txt)

**Comment by Tnnnguyen** [comment]
Using the fix with the wait: [here](https://github.com/smartdevicelink/sdl_android/tree/bugfix/usb_enhacements_wait):
Both phones appear to connect consistently at least 10 times in a row.  I ran 2 sets for each phone, killing the app and rebooting the TDK between sets.

**Comment by mrapitis** [comment]
Initial test results on newer devices look good so far.  It seems like some older devices may be  having trouble with the enterForeground call.  
```
I/dalvikvm(31005): Could not find method android.app.Notification$Builder.setChronometerCountDown, referenced from method com.smartdevicelink.transport.SdlRouterService.enterForeground
W/dalvikvm(31005): VFY: unable to resolve virtual method 221: Landroid/app/Notification$Builder;.setChronometerCountDown (Z)Landroid/app/Notification$Builder;
D/dalvikvm(31005): VFY: replacing opcode 0x6e at 0x013b
I/dalvikvm(31005): Could not find method android.app.Notification$Builder.<init>, referenced from method com.smartdevicelink.transport.SdlRouterService.enterForeground
W/dalvikvm(31005): VFY: unable to resolve direct method 212: Landroid/app/Notification$Builder;.<init> (Landroid/content/Context;Ljava/lang/String;)V
D/dalvikvm(31005): VFY: replacing opcode 0x70 at 0x006c
E/dalvikvm(31005): Could not find class 'android.app.NotificationChannel', referenced from method com.smartdevicelink.transport.SdlRouterService.enterForeground
W/dalvikvm(31005): VFY: unable to resolve new-instance 53 (Landroid/app/NotificationChannel;) in Lcom/smartdevicelink/transport/SdlRouterService;
D/dalvikvm(31005): VFY: replacing opcode 0x22 at 0x017b
I/SELinux (31037): Function: selinux_android_load_priority , spota verifySig and checkHash pass. priority version is VE=SEPF_GT-N7100_4.4.2_0040
I/SELinux (31037):  
```
Attaching log file for details.
[crashlogs.txt](https://github.com/smartdevicelink/sdl_android/files/2648986/crashlogs.txt)

**Comment by mrapitis** [comment]
@joeygrover in testing we have seen improvements in connectivity across many devices, however there are still some devices that we have noticed go into an unrecoverable state where the app can only be discovered after a force close.  We will continue testing and post more feedback as results become available.

**Comment by hypatankar** [comment]
@joeygrover If one disconnects the phone when the phone is connecting to the TDK (specifically the instant when the white screen pops up), the phone will enter a state where it says it is connected via USB when it has not.  Assuming one does not restart phone, TDK, or force close helloSDL, it is not possible to recover from this and get the helloSDL app to register on the TDK. We have noticed this on multiple devices including Google Pixel and Moto Z Force Droid. Logs from Moto Z Force Droid running Android 7.1.1 are attached.

[moto z breaking after disconnect when connecting.txt](https://github.com/smartdevicelink/sdl_android/files/2664911/moto.z.breaking.after.disconnect.when.connecting.txt)

Reproduction steps are as follow:
1)  Build the Hello SDL sample application included with the SDL Android proxy library and install it on the Android mobile device.
2)  Stand up an instance of SDL core 5.0.0 on Linux or acquire hardware such as a Sync TDK that has AOA USB transport support.
3) Physically connect the USB cable to the TDK or PC running SDL core.
4) Set app as default app for USB connection when dialog displays
5) Take note that the Hello SDL sample app generally successfully registers with SDL core.
6) Physically disconnect the USB cable from the TDK or PC.
7) Physically connect the USB cable to the TDK or PC running SDL core.
8) Immediately unplug phone before it finishes connecting when the white screen first  appears on phone
9) Repeat steps 6 and 7 with gap of 5 seconds between them, helloSDL should not register

**Comment by joeygrover** [comment]
@hypatankar Yes this relates to the fact that Android will let the app open the USB accessory even when the USB cord is not physically attached. Circumenting this might require some very hacky code.

**Comment by jordynmackool** [comment]
Fords testing document is attached from the SC meeting on 2018-12-18:
[Attachement 1- USB AOA Testing.xlsx](https://github.com/smartdevicelink/sdl_android/files/2764657/Attachement.1-.USB.AOA.Testing.xlsx)

**Comment by joeygrover** [comment]
@mrapitis @hypatankar the PR has been updated with additional protections against the known fail case of USB disconnect during activity on screen. Also took care of issue with app service not shutting down during disconnect (StartService sent, but ACK never received) which I believe is one of the nonrecoverable states. Please give the update PR a round of testing.

**Comment by mrapitis** [comment]
@joeygrover we will test with the additional changes and keep you posted.  Thanks for the update!

**Comment by mrapitis** [comment]
@jordynmackool please find attached test results. Thank you.
[USB AOA Testing.xlsx](https://github.com/smartdevicelink/sdl_android/files/2836405/USB.AOA.Testing.xlsx)

**Comment by jordynmackool** [comment]
@mrapitis Can you please share a summary on the test results and provide the log files that are linked in the test document?

**Comment by mrapitis** [comment]
@jordynmackool 

Overall test results show AOA connectivity has significantly improved over the current develop branch; in some cases connectivity has even improved over the pre 4.7.x releases.  We do recommend including these changes in the upcoming 4.8 release.  Logs are also attached.
thanks
[Logs.zip](https://github.com/smartdevicelink/sdl_android/files/2852063/Logs.zip)

**Comment by jordynmackool** [comment]
It was determined on 2019-02-12 in the Steering Committee Meeting that this issue can be closed and the PR fix will be included in the upcoming 4.8 proxy release.

**Comment by theresalech** [comment]
This issue has been resolved in the 4.8 release.
